### PR TITLE
refactor(NODE-4144): remove all symbol properties

### PR DESCRIPTION
**Pull request by nbbeeken**
### Description

Similar to: https://github.com/mongodb/node-mongodb-native/pull/4133

#### What is changing?

- Remove all symbol class properties in favor of private and internal marked string properties

##### Is there new documentation needed for these changes?

No

#### What is the motivation for this change?

Symbol class properties were introduced to the driver before we adopted typescript and API-extractor to clearly indicate the accessibility of driver APIs. The symbol properties provided implied privacy, making properties non-trivial to access. Today we have the power of typescript private/protected and API-extractor's type deletion feature for APIs marked internal. With these tools, we can remove symbol properties in favor of appropriately marked string properties. This makes debugging internals easier by making inspection and assertion far more simple.

### Double check the following

- [ ] Ran `npm run check:lint` script
- [ ] Self-review completed using the [steps outlined here](https://github.com/mongodb/node-mongodb-native/blob/HEAD/CONTRIBUTING.md#reviewer-guidelines)
- [ ] PR title follows the [correct format](https://www.conventionalcommits.org/en/v1.0.0/): `type(NODE-xxxx)[!]: description`
  - Example: `feat(NODE-1234)!: rewriting everything in coffeescript`
- [ ] Changes are covered by tests
- [ ] New TODOs have a related JIRA ticket

### CODE REVIEW

File: src/encrypter.ts [hunk 1]
```diff
@@ -96,7 +93,7 @@ export class Encrypter {
 
   async connectInternalClient(): Promise<void> {
     // TODO(NODE-4144): Remove new variable for type narrowing
```
Comment (baileympearson) [inline src/encrypter.ts:95]: This comment should no longer be relevant, right?

File: src/error.ts [hunk 2]
```diff
@@ -128,8 +125,7 @@ function isAggregateError(e: unknown): e is Error & { errors: Error[] } {
  * mongodb-client-encryption has a dependency on this error, it uses the constructor with a string argument
  */
 export class MongoError extends Error {
-  /** @internal */
-  [kErrorLabels]: Set<string>;
+  public readonly errorLabels: string[] = [];
```
[the remark this answers is not in the document]
Comment (baileympearson) [reply src/error.ts:128]: What indirection are you referring to here?  That it was private or that it was a set of strings?

File: src/error.ts [hunk 3]
```diff
@@ -188,15 +183,11 @@ export class MongoError extends Error {
    * @returns returns true if the error has the provided error label
    */
   hasErrorLabel(label: string): boolean {
-    return this[kErrorLabels].has(label);
+    return this.errorLabels.includes(label);
```
[the remark this answers is not in the document]
Comment (baileympearson) [reply src/error.ts:186]: I would be surprised if the set were slower than the work you're doing here - Set is likely optimized by V8, but you've essentially reproduced the same behavior with extra steps (you iterate over all error labels, and for each one you check if the array includes the label (O(n))).

Why do you think this is faster than using a set?

File: src/error.ts [hunk 4]
```diff
@@ -1065,13 +1051,18 @@ export class MongoNetworkError extends MongoError {
...
   }
 
   override get name(): string {
     return 'MongoNetworkError';
   }
+
+  /** @internal */
+  static isBeforeHandshake(err: MongoNetworkError): boolean {
+    return err.beforeHandshake === true;
+  }
```
[the remark this answers is not in the document]
Comment (baileympearson) [reply src/error.ts:1065]: Any reason this isn't an instance method?

File: src/mongo_client.ts [hunk 5]
```diff
@@ -367,20 +364,22 @@ export class MongoClient extends TypedEventEmitter<MongoClientEvents> implements
 
   /**
    * The consolidate, parsed, transformed and merged options.
-   * @internal
    */
-  [kOptions]: MongoOptions;
+  public readonly options: Readonly<
+    Omit<MongoOptions, 'monitorCommands' | 'ca' | 'crl' | 'key' | 'cert'>
+  > &
+    Pick<MongoOptions, 'monitorCommands' | 'ca' | 'crl' | 'key' | 'cert'>;
```
[the remark this answers is not in the document]
Comment (baileympearson) [reply src/mongo_client.ts:371]: Do we mutate `'monitorCommands' | 'ca' | 'crl' | 'key' | 'cert'`?

File: src/sessions.ts [hunk 6]
```diff
@@ -184,12 +175,14 @@ export class ClientSession
...
     if (options.snapshot === true) {
-      this[kSnapshotEnabled] = true;
+      this.snapshotEnabled = true;
       if (options.causalConsistency === true) {
         throw new MongoInvalidArgumentError(
           'Properties "causalConsistency" and "snapshot" are mutually exclusive'
         );
       }
+    } else {
+      this.snapshotEnabled = false;
```
[the remark this answers is not in the document]
Comment (baileympearson) [reply src/sessions.ts:185]: (optional) we can flatten this if we want:

```typescript
this.snapshotEnabled = options.snapshot === true;
if (this.snapshotEnabled && options.causalConsistency === true) {
  throw new MongoInvalidArgumentError(
    'Properties "causalConsistency" and "snapshot" are mutually exclusive'
   );
}
```

**Review by baileympearson — CHANGES_REQUESTED** [review]

File: src/error.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (baileympearson) [reply src/error.ts:1065]: Actually, this is just checking a boolean value.  Do we need this method anymore at all?

File: src/error.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (nbbeeken) [reply src/error.ts]: Fair points made, I'm over simplifying here, reverted.

File: src/mongo_client.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (nbbeeken) [reply src/mongo_client.ts:371]: We do! There's a getter/setter for monitorCommands and those TLS values are set after `.connect` is called.

File: src/error.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (nbbeeken) [reply src/error.ts]: Undid this.

File: src/sessions.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (nbbeeken) [reply src/sessions.ts]: sure!

File: src/change_stream.ts [hunk 7]
```diff
@@ -806,7 +794,7 @@ export class ChangeStream<
   }
 
   async *[Symbol.asyncIterator](): AsyncGenerator<TChange, void, void> {
-    if (this.closed) {
+    if (this.isClosed) {
```
[the remark this answers is not in the document]
Comment (baileympearson) [reply src/change_stream.ts:797]: Sorry, that wasn't clear.  I wasn't suggesting using the getter for the sake of using a getter.  By using `.isClosed`, you have changed the logic.  The getter is defined as:

```typescript
  /** Is the cursor closed */
  public get closed(): boolean {
    return this.isClosed || this.cursor.closed;
  }
```

Was removing the check for `this.cursor.closed` intentional?

**Review by baileympearson — CHANGES_REQUESTED** [review]

File: src/change_stream.ts (hunk 7 above)
[the remark this answers is not in the document]
Comment (nbbeeken) [reply src/change_stream.ts]: Oh good catch, fixed!

File: src/change_stream.ts [hunk 8]
```diff
@@ -855,7 +843,7 @@ export class ChangeStream<
    * @throws MongoChangeStreamError if the underlying cursor or the change stream is closed
    */
   stream(options?: CursorStreamOptions): Readable & AsyncIterable<TChange> {
-    if (this.closed) {
+    if (this.isClosed) {
```
[the remark this answers is not in the document]
Comment (nbbeeken) [reply src/change_stream.ts]: also fixed!

**Review by baileympearson — DISMISSED** [review]

**Review by baileympearson — APPROVED** [review]

**Pull request merged**